internal/suite/setup.go

```diff
@@ -0,0 +1,96 @@
+package suite
+
+import (
+	"context"
+	"os"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	cr "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/clustertest"
+	"github.com/giantswarm/clustertest/pkg/application"
+	"github.com/giantswarm/clustertest/pkg/client"
+	"github.com/giantswarm/clustertest/pkg/logger"
+	"github.com/giantswarm/clustertest/pkg/wait"
+
+	"github.com/giantswarm/cluster-test-suites/internal/state"
+)
+
+// ClusterBuilder is an interface that provides a function for building provider-specific Cluster apps
+type ClusterBuilder interface {
+	NewClusterApp(clusterName string, orgName string, clusterValuesFile string, defaultAppsValuesFile string) *application.Cluster
+}
+
+// Setup handles the creation of the BeforeSuite and AfterSuite handlers. This covers the creations and cleanup of the test cluster.
+// `clusterReadyFns` can be provided if the cluster requires custom checks for cluster-ready status. If not provided the cluster will
+// be checked for at least a single control plane node being marked as ready.
+func Setup(isUpgrade bool, kubeContext string, clusterBuilder ClusterBuilder, clusterReadyFns ...func(client *client.Client)) {
```

Comment:
Just a thought: you can just pass the NewClusterApp function to Setup instead of the interface so you don't have to create so many new types and then pass empty objects around just to use the one function (unless we see the interface and types growing in the future).

Reply:
Or even just pass the provider and let Setup decide which library to call out to.

Reply:
Wont that lead to cyclical imports?

Reply:
Actually, it wont thinking about it. But I'm not sure the provider would work as we need different logic for CAPA with private networking, for example.

Reply:
> But I'm not sure the provider would work as we need different logic for CAPA with private networking, for example.

We can then use the KubeContext buuut then we have to store our own mapping so I don't like it anymore :D

> That was my thinking. Not sure what that might be yet. Maybe nothing. What do you think?

I can't currently think of anything either but who knows. Happy to keep it this way as we have it implemented already then :)

Reply:
We could always refactor it to just pass a function at a later date if we decide thats enough. :)